### PR TITLE
HPCC-17834 Alternate URL pages should redirect to Login.html

### PR DESCRIPTION
**Pull request by kunalaswani**
- Enforces redirect to the login page
- Bookmarked pages will now redirect to login page when user is logged out

Signed-off-by: Aswani, Kunal <kunal.aswani@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

### CODE REVIEW

**Comment by kunalaswani** [comment]
@miguelvazq  Please review, lws.config.js file included due to updates

File: esp/src/stub.htm [hunk 1]
```diff
@@ -34,6 +34,15 @@
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/node_modules.js\" charset=\"utf-8\"><" + "/script>");
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/stub.eclwatch.js\" charset=\"utf-8\"><" + "/script>");
         }
+
+        checkCookie ();
+        function checkCookie(){
+            if (document.cookie.indexOf ("ESPAuthenticated") > -1){
+                window.location;
```
Comment (miguelvazq) [inline esp/src/stub.htm]: Isn't `window.location.reload();` what your looking for?

File: esp/src/eclwatch/stub.js [hunk 2]
```diff
@@ -214,4 +214,4 @@ define([
         parseUrl();
         initUI();
     });
-});
+});
```
Comment (miguelvazq) [inline esp/src/eclwatch/stub.js]: Why remove the line?

File: esp/src/eclwatch/LockDialogWidget.js [hunk 3]
```diff
@@ -119,7 +120,7 @@ define([
 
             xhr("esp/lock", {
                 method: "post",
-            });
```
Comment (miguelvazq) [inline esp/src/eclwatch/LockDialogWidget.js]: Cookie not used? Why remove the `;`?

**Review by miguelvazq — COMMENTED** [review]
Some comments.

File: esp/src/stub.htm (hunk 1 above)
Comment (kunalaswani) [reply esp/src/stub.htm]: No, when using window.location.reload(); the checkCookie process gets stuck in an endless loop that never lets stub load. I will change this to window.location();

**Comment by kunalaswani** [comment]
Suggested changes made and updated, please review @miguelvazq

File: esp/src/stub.htm [hunk 4]
```diff
@@ -34,6 +34,15 @@
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/node_modules.js\" charset=\"utf-8\"><" + "/script>");
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/stub.eclwatch.js\" charset=\"utf-8\"><" + "/script>");
         }
+
+        checkCookie ();
+        function checkCookie(){
+            if (document.cookie.indexOf ("ESPAuthenticated") > -1){
```
Comment (miguelvazq) [inline esp/src/stub.htm]: I would write it like this:
`document.cookie.indexOf("ESPAuthenticated") > -1 ?  window.location() : window.location = (location.origin + "/esp/files/Login.html");`

**Comment by kunalaswani** [comment]
Tested, Built, and checked. Please review

File: esp/src/stub.htm [hunk 5]
```diff
@@ -34,6 +34,11 @@
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/node_modules.js\" charset=\"utf-8\"><" + "/script>");
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/stub.eclwatch.js\" charset=\"utf-8\"><" + "/script>");
         }
+
+        checkCookie ();
+        function checkCookie(){
+            if (document.cookie.indexOf ("ESPAuthenticated") > -1 ? window.location() :window.location = location.origin + "/esp/files/Login.html");
```
Comment (miguelvazq) [inline esp/src/stub.htm]: The `if()` is not needed in this case

**Comment by AttilaVamos** [comment]
@richardkchapman Roxie port (9876) problem again. Do you have any idea?

**Comment by mckellyln** [comment]
Hi.  I'll take a look.  What machine does it run on ?

**Comment by AttilaVamos** [comment]
It is already wiped out. (BTW: 10.241.40.11 and actually PR-11228 builds on it)

File: esp/src/stub.htm [hunk 6]
```diff
@@ -34,6 +34,11 @@
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/node_modules.js\" charset=\"utf-8\"><" + "/script>");
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/stub.eclwatch.js\" charset=\"utf-8\"><" + "/script>");
         }
+
```
Comment (miguelvazq) [inline esp/src/stub.htm]: Minor: Remove extra space.

**Review by miguelvazq — COMMENTED** [review]
Last change before I test.

File: esp/src/stub.htm (hunk 6 above)
Comment (kunalaswani) [reply esp/src/stub.htm]: removed

**Comment by richardkchapman** [comment]
@miguelvazq Back to you

**Comment by kunalaswani** [comment]
@richardkchapman please hold off on this merge, i am testing non session based environments to make sure it does not affect them.

**Comment by richardkchapman** [comment]
@kunalaswani Back to you

**Comment by kunalaswani** [comment]
@miguelvazq please review updated redirect code

File: esp/src/stub.htm [hunk 7]
```diff
@@ -34,6 +34,42 @@
...
+            } else if (document.cookie.indexOf ("ESPSessionState=false") > -1) {
+                window.location();
+            } else {
+                sendAuthRequest();
+                    function sendAuthRequest(){
+                        var xhttp = new XMLHttpRequest();
+                        xhttp.open ('GET', 'esp/getauthtype', true);
+                        xhttp.send();
+                        xhttp.onload = function(){
+                            if (this.readyState == 4 && this.status == 200 && this.responseXML.childNodes[0].textContent == "Mixed") {
```
Comment (miguelvazq) [inline esp/src/stub.htm]: Repetitive code lets try to shorten this by using something like a switch.

**Comment by kunalaswani** [comment]
@miguelvazq please review with new switch logic.

File: esp/src/eclwatch/HPCCPlatformWidget.js [hunk 8]
```diff
@@ -462,6 +462,7 @@ define([
                         if (data) {
                             cookie("ECLWatchUser", "", { expires: -1 });
                             cookie("ESPSessionID" + location.port + " = '' ", "", { expires: -1 });
+                            cookie("ESPAuthenticated", "", { expires: -1 });
```
Comment (miguelvazq) [inline esp/src/eclwatch/HPCCPlatformWidget.js]: Is this still needed?

File: esp/src/stub.htm [hunk 9]
```diff
@@ -34,6 +34,43 @@
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/node_modules.js\" charset=\"utf-8\"><" + "/script>");
             document.write("<script type=\"text/javascript\" src=\"/esp/files/dist/stub.eclwatch.js\" charset=\"utf-8\"><" + "/script>");
         }
+        getAuthType();
```
Comment (miguelvazq) [inline esp/src/stub.htm]: How about something like this? 

       getAuthType();
           
       function getAuthType(){
            if (document.cookie.indexOf ("ESPSessionState=true") > -1){
                checkCookie();
            } else if (document.cookie.indexOf ("ESPSessionState=false") > -1) {
                window.location();
            } else {
                sendAuthRequest();
            }
        }

        function checkCookie(){
            document.cookie.indexOf ("ESPAuthenticated=true") > -1 ? window.location() :window.location = location.origin + "/esp/files/Login.html";
        }

        function sendAuthRequest(){
            var xhttp = new XMLHttpRequest();
            xhttp.open ('GET', 'esp/getauthtype', true);
            xhttp.send();
            xhttp.onload = function(){
                var responseType = (this.responseXML.childNodes[0].textContent);
                if (this.readyState == 4 && this.status == 200 && this.responseXML.childNodes[0].textContent == responseType){
                    switch (responseType) {
                        case 'Mixed':
                        case 'PerSessionOnly':
                        case 'PerRequestOnly':
                            document.cookie = ("ESPSessionState=true");
                        break;
                        case 'UserNameOnly':
                        case 'None':
                            document.cookie = ("ESPSessionState=false");
                        break;
                    }
                } else {
                    alert ('Authorization Request Error.')
                }
            }
        }

**Review by miguelvazq — COMMENTED** [review]
See my recommendation. Will need a re-test.

**Comment by kunalaswani** [comment]
@miguelvazq removed ("cookie("ESPAuthenticated", "", { expires: -1 });") from HPCCPlatformWidget.js, tested, and no longer needed. Also rewrote the switch logic, with your suggestions, tested against the options that we offer and appears to function as expected. Please review.

**Comment by miguelvazq** [comment]
Looks good.

**Comment by kunalaswani** [comment]
@richardkchapman this ticket is dependent on https://github.com/wangkx/HPCC-Platform/commit/2348c5801c119b3e81067808569dc1188283737d and should be merged only after @wangkx 's ticket has been merged.

**Comment by kunalaswani** [comment]
This was retested after @wangkx 's pull request was merged and is working as expected. It should be safe to merge @richardkchapman

**Comment by richardkchapman** [comment]
@miguelvazq You ok with this being merged now?

**Comment by miguelvazq** [comment]
Reviewed again. Looks fine.

**Pull request merged**